### ISSUE TITLE
Wrong cast generated for Informix and others when casting a LocalDateTime value

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#8561]
The following query produces erroneous SQL in Informix.

```java
System.out.println(ctx.selectOne()
    .where(currentLocalDateTime().gt(LocalDateTime.of(2000, 1, 1, 0, 0, 0))).fetch());
```

The bad SQL being produced is this:

```sql
select 1 "one" 
from (select 1 as dual from systables where tabid = 1) as dual 
where current > cast(? as timestamp)
```

Curiously, when inlining bind values, the generated syntax is correct:

```sql
select 1 "one" 
from (select 1 as dual from systables where tabid = 1) as dual 
where current > datetime(2000-01-01 00:00:00.0) year to fraction
```

This can be worked around by preventing casts around bind variables by applying:

```java
settings.setParamCastMode(ParamCastMode.NEVER);
```

**Comment by lukaseder** [comment]
The problem does not appear with JDBC `Timestamp` types:

```java
System.out.println(ctx.selectOne()
    .where(currentTimestamp().gt(Timestamp.valueOf(LocalDateTime.of(2000, 1, 1, 0, 0, 0)))).fetch());
```

**Comment by lukaseder** [comment]
It appears that for Informix, the `SQLDataType.LOCALDATETIME` type is not registered in the internal `DefaultDataType.TYPES_BY_SQL_DATATYPE` registry.

**Comment by lukaseder** [comment]
This is the case for all dialects, not just Informix. Informix seems to be the only one so far that runs into this problem because:

- It always casts bind variables to their type
- Its `timestamp` type is called `datetime`

**Comment by lukaseder** [comment]
I can confirm this also happens for SQL Server and it is not related to bind variable casting but to casting in general. This also fails in SQL Server:

```java
Record3<LocalDate, LocalTime, LocalDateTime> result =
ctx.fetchSingle(
    val(d).cast(LocalDate.class), val(t).cast(LocalTime.class), val(dt).cast(LocalDateTime.class)
);

assertEquals(d, result.value1());
assertEquals(t, result.value2());
assertEquals(dt, result.value3());
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.12. Backport scheduled for 3.11.12 (#8562)
